File: prototype/spellcheck.py
#!/bin/env python

""" spellcheck for dictionary, run when input not found """

from util import get_word_list

class SpellCheck(object):
	""" Spellcheck word based on self.vocabulary """
	def __init__(self, source):
		self.vocabulary = set(source)

	def candidates(self, word):
		""" Generate possible spelling corrections for word. """
		return (self._known([word])
				or self._known(self._edits1(word))
				or self._known(self._edits2(word))
				or [word])

	def _known(self, words):
		""" The subset of `words` that appear in the vocabulary """
		return set(w for w in words if w in self.vocabulary)

	def _edits1(self, word):
		""" All edits that are one edit away from `word`. """
		letters   = 'abcdefghijklmnopqrstuvwxyz'
		splits    = [(word[:i], word[i:])    for i in range(len(word) + 1)]
		deletes   = [L + R[1:]               for L, R in splits if R]
		swap      = [L + R[1] + R[0] + R[2:] for L, R in splits if len(R) > 1]
		replaces  = [L + c + R[1:]           for L, R in splits if R for c in letters]
		inserts   = [L + c + R               for L, R in splits for c in letters]
		return (deletes + swap + replaces + inserts)

	def _edits2(self, word):
		""" All edits that are two edits away from `word`. """
		return (e2 for e1 in self._edits1(word) for e2 in self._edits1(e1))

def main():
	""" main function """
	english_words = get_word_list()
	spellcheck = SpellCheck(english_words)

	while True:
		wrong_word = input('Enter a wrong word: ')
		print('Candidates:')
		candidates = spellcheck.candidates(wrong_word)
		for candidate in candidates:
			print(candidate)

if __name__ == '__main__':
	main()

# vim: nofoldenable

File: src/EDictionary.Core/Utilities/SpellCheck.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;

namespace EDictionary.Core.Utilities
{
	static public class SpellCheck
	{
		public static HashSet<string> Vocabulary { get; set; }
		private const string alphabet = "abcdefghijklmnopqrstuvwxyz";
		private static Func<string, bool> IsValidWord = word => Vocabulary.Contains(word);

		/// <summary>
		/// initialize Vocabulary for SpellCheck class
		/// </summary>
		public static void GetVocabulary(IEnumerable<string> wordList)
		{
			Vocabulary = new HashSet<string>(wordList);

			foreach (var word in wordList)
				Vocabulary.Add(word.ToLower());
		}

		/// <summary>
		/// Return an IEnumarable of words (valid or not) that are one edit away from the given word
		/// <para/>
		/// Example: 'strike' would return {'trike', 'tsrike', 'atrike', 'astrike', ...}
		/// </summary>
		public static IEnumerable<string> Edits1(string word)
		{
			var splits = from i in Enumerable.Range(0, word.Length)
				select new {a = word.Substring(0, i), b = word.Substring(i)};

			var deletes = from s in splits
				where s.b != "" // we know it can't be null
				select s.a + s.b.Substring(1);

			var transposes = from s in splits
				where s.b.Length > 1
				select s.a + s.b[1] + s.b[0] + s.b.Substring(2);

			var replaces = from s in splits
				from c in alphabet
				where s.b != ""
				select s.a + c + s.b.Substring(1);

			var inserts = from s in splits
				from c in alphabet
				select s.a + c + s.b;

			return deletes
				.Union(transposes) // union translates into a set
				.Union(replaces)
				.Union(inserts);
		}

		/// <summary>
		/// Return an IEnumarable of words that are two edits away from the given word
		/// See Edits1(string)
		/// </summary>
		private static IEnumerable<string> Edits2(string word)
		{
			return (
					from e1 in Edits1(word)
					from e2 in Edits1(e1)
					select e2
					).Distinct();
		}

		/// <summary>
		/// return a list of valid words from another wordlist
		/// </summary>
		private static IEnumerable<string> Known(IEnumerable<string> words)
		{
			return words.Where(word => IsValidWord(word));
		}

		/// <summary>
		/// return a list of candidates for the wrong spelling word
		/// </summary>
		public static IEnumerable<string> Candidates(string word)
		{
			var candidates = new[] {
					Known(new[] {word}),
					Known(Edits1(word)),
					Known(Edits2(word)),
					new[] {word},
					}.First(knowns => knowns.Any());

			return candidates;
		}

		public static void ReadFromStdIn()
		{
			string word;

			while (!string.IsNullOrEmpty(word = Console.ReadLine().Trim()))
			{
				foreach (var candidate in Candidates(word))
				{
					Console.Write(candidate + " ");
				}
				Console.WriteLine();
			}
		}
	}
}
